src/gccrs/args.rs

```diff
@@ -0,0 +1,152 @@
+        options.optmulti("", "crate-type", "Type of binary to output", "TYPE");
+
+        options
+    }
+
+    fn format_output_filename(matches: &Matches, crate_type: &CrateType) -> String {
+        // FIXME: No unwraps here
+        let crate_name = matches.opt_str("crate-name").unwrap();
+        let out_dir = matches.opt_str("out-dir").unwrap();
+
+        // FIXME: Figure out a way to return multiple output filenames. Just handle `bin`
+        // for now
+        let c_options = matches.opt_strs("C");
+
+        // FIXME: This is probably different on Windows, we should use Paths and PathStrs
+        // instead
+        let mut output_file = match crate_type {
+            &CrateType::Bin => format!("{}/{}", out_dir, crate_name),
+            &CrateType::DyLib => format!("{}/lib{}.so", out_dir, crate_name),
+            &CrateType::StaticLib => format!("{}/lib{}.a", out_dir, crate_name),
+            _ => unreachable!(
+                "gccrs cannot handle other crate types than bin, dylib or staticlib at the moment"
+            ),
+        };
+
+        // FIXME: Horrendous. We need to create a separate "C options" parser since we'll
+        // probably use more than just `extra-filename`.
+        c_options
+            .iter()
+            .for_each(|c_opt| match c_opt.split('=').nth(0).as_deref() {
+                Some("extra-filename") => output_file.push_str(&c_opt.split('=').nth(1).unwrap()),
+                _ => {}
+            });
+
+        dbg!(output_file)
+    }
+
+    /// Get the corresponding `gccrs` argument from a given `rustc` argument
+    pub fn from_rustc_args(rustc_args: &Vec<String>) -> Vec<GccrsArgs> {
+        let options = GccrsArgs::generate_parser();
+
+        // Parse arguments, skipping `cargo-gccrs` and `rustc` in the invocation
+        let matches = match options.parse(&rustc_args[2..]) {
+            Ok(m) => m,
+            Err(err) => unreachable!("{:?}", err),
+        };
+
+        matches
+            .opt_strs("crate-type")
+            .iter()
+            .map(|type_str| CrateType::from_str(&type_str))
+            .map(|crate_type| {
+                (
+                    // We need to keep the crate_type as well, in order to spawn the
+                    // correct command. Return it in a tuple alongside the generated
+                    // output filename.
+                    crate_type,
+                    GccrsArgs::format_output_filename(&matches, &crate_type),
+                )
+            })
+            .map(|(crate_type, output_file)| GccrsArgs {
+                source_files: matches.free.clone(),
+                crate_type,
+                output_file,
+            })
+            .collect()
+    }
+
+    /// Convert a `GccrsArgs` structure into arguments usable to spawn a process
+    pub fn to_args(mut self) -> Vec<String> {
+        let mut args = vec![];
+
+        // Add all the source files to the command line
+        self.source_files
+            .drain(..)
```

Comment:
Why do you have to `drain` here? This is effectively just moving the contests of `source_files` into `args`. Since you moved `self` into this function anyway, you could also just use `source_files` directly. Or just do this:

```rust
let mut args = self.source_files;
```